### ISSUE TITLE
🚨 Prerelease feedback: v0.2.2-preview.0

### ISSUE THREAD

**Issue by croian** [croian/mouseless-issues#79]
UPDATE, Dec 21: [v0.2.2-preview.1](https://github.com/croian/mouseless-issues/releases/tag/v0.2.2-preview.1) has been uploaded

Any help testing out prereleases is super appreciated 🙏 and gives you early access to features! 🚀

I will set up an official prerelease channel (or channels) in the near future, but I figured this was the best place for notifying everyone and gathering feedback for now.

I just uploaded the .dmg to [v0.2.2-preview.0](https://github.com/croian/mouseless-issues/releases/tag/v0.2.2-preview.0).

So far, it contains fixes for:
https://github.com/croian/mouseless-issues/issues/39 (Non-english layouts/characters not working)
https://github.com/croian/mouseless-issues/issues/30 (Some chars on alternate english layouts (e.g. dvorak) not working)
https://github.com/croian/mouseless-issues/issues/52 (minor: updated some menu items' text)

Thank you so much to anyone who can test and give feedback **before/by ~~Friday~~ Monday** -- it gives me extra confidence in the release, and I really appreciate it!! 🙏

**Comment by twiggshub** [comment]
I don't know if I'm the only one with this problem so far but I needed to do a complete re-installation including removing the folders /Mouseless/Data to launch the app after dragging the preview .app into my applications folder. 

Could it have something to do with the data folder not updating when replacing the .app file or something? To fix it I 

* Removed the data folder
* Put in the DMG again
* Reset the config.yaml to the default one that comes with the new pre-release.

**Comment by twiggshub** [comment]
Yes, of course! 

Error messages, though it still runs as a process in the background (although it can not be shut down). No popups, nor status bar icons show up. Resetting the config did not work either, except deleting the data folder, running the app again and letting it build everything. 

Other than that I sadly don't know what to look at specifically.

**Comment by croian** [comment]
@twiggshub thank you for the extra info!  a few questions to drill down on this:

- would you mind screenshotting the error message(s) you're seeing?
- is mouseless running when you drag the new .app file into the Applications folder?  in my testing, trying to do so while it is running simply informs you that the operation can't be done.
- can you confirm you are following this process exactly?
  - open the .dmg file
  - drag the .app file to the Applications folder
  - choose "Replace" instead of "Keep both"
  - start Mouseless

your help diagnosing this is most appreciated 🙏 could help prevent any issues for a lot of users in the future

**Comment by vlnn** [comment]
_(I had no problems with reinstalling Mouseless, except MacOS asked anew if I'm ready to run some app downloaded from the Internet.)_

My test report regarding fixes on #39 : the behavior has been changed and is much better, but some problems are still there. 

### New and expected behavior:

Now I am able to add Cyrillic characters to the `Mouse action keys` as well as to `Sub-grid mouse action keys` and they are correctly working when my keyboard is set to Ukrainian layout — e.g. I can access the cell marked ЙЙ instead of QQ.

## New and non-expected behavior:
1. Even though the fix provides a general solution for non-english based keyboard layouts, there's no way to change the layout when the Mouseless' popup is active, i.e. pressing `Globe/Fn` button is not changing the layout, so if the popup was invoked while being in `EN` layout, user can't input anything from `UA`. The indicator of keyboard layout in the menu bar is not changing under the Mouseless popup as well.
2. Even with (future) fix to previous problem Mouseless would be hard to use for multi-layout setups. The problem here is that most of the multi-language users are using two layouts constantly changing the layout, and also layout is automatically changes when you switch apps (so I can have different layouts "bound" to different windows — there's a configuration option for that **Automatically switch to a document's input source**).
So multi-language users instead of having "happy path" of "invoke Mouseless, type two chars, type space" will be bound to "not-that-happy path" of "invoke Mouseless, type two chars, have no result as the input mode is set to `EN`, hesitate, change the language to UA, type two chars again". What is worse, this issue will be "flakey" as it depends on transient current input mode.

There are two usual (or rather known to me) ways to resolve this problem in programs I'm using: either create an (user) mapping each "standard character" to "Cyrillic character" (e.g. `QЙ`,`WЦ`,`EУ`) — which is quite some configuration, but done just once for each program ("vim" and "vimperator" and "Pentadactyl" use this way), or just ignore input mode as a whole and use key codes instead of real characters (this is how https://github.com/televator-apps/vimari works AFAIK).

@croian, Please let me know, if additional information is needed: this is crucial issue for me and millions of people using more than one layout, so I'm ready to invest some time into the testing and discussions. I also hope that my wall of text does not dishearten you — this is due to scenario that is unusual to `EN`-only typers.

**Comment by croian** [comment]
hi @vlnn the detailed response is much appreciated!  particularly since I'm a qwerty-only user myself

i think a quick and easy solution to alleviate 1) is to simply have a special case for the globe/fn key, and allow it to pass through to the system while the overlay is up.  not a perfect solution, as some users use different shortcuts to switch layouts (i've been using ctrl+space while trying to fix this issue, because i don't have a fn key on my keyboard), and because, as you mention, we ultimately have issue 2) to contend with.

- the ultimate solution, i think would indeed be to have the app automatically recognize layouts and map chars to keycodes and vice versa, without needing any configuration or effort by the user.  i think there are macos APIs for this, but it might take some extra legwork to get them working with mouseless, but I will a path for doing so ASAP

- another solution in the meantime, would be to have an interface for the user enter all the keys for their different layouts, and map them to keycodes.  it wouldn't take too long, but still might annoy and confuse users as to why its necessary. but if that's what it takes to solve the issue soon it might be the best option for now.
  - you mention vim/vimperator/pentadactyl have some sort of configuration like this -- do they actually require the user to map characters to characters?  or do they use something like the approach i mentioned above?

- finally, one of the most fragile alleviations would be to try to get mouseless to actually utilize the "automatically switch to document's input source" feature, but it might involve rapidly changing app/window focuses between mouseless and the previously focused window, which sounds very janky
  - just so i understand the problem space better though -- if i did get this approach to work, would that mostly solve #2 for you, or anyone, who has the "automatically switch" option enabled?

please let me know if you have any clarifications or corrections to my understanding of the situation.  thank you again for the detailed information!

**Comment by twiggshub** [comment]
Wish I could provide a screenshot of the error message, but none shows up. There is no error message. No icon in the status bar, however the process itself was running and appeared in the activity monitor as a killable process. Is there any other info I can provide on this?

Second: Mouseless, at the time of the issue was not running. Of course you can’t delete an app or replace an app while it’s still running. 

Third: This process is correct. 

Also, come to think about it, I tried replicating the issue before I posted this, figured it probably was something to do with the .config file before I deleted the entire data folder. AFTER replacing the .dmg file, my config.yaml file still had the incorrect version "app_version: 0.2.2-preview.0” number on it (Don’t know if it’s completely unrelated reason why).  

Let me know if there’s anything I could provide

**Comment by croian** [comment]
@twiggshub thank you for the update -- i think i'm understanding the problem more clearly now, and it could be a parsing error or some other issue with the config that just isn't handled correctly.  i tried playing with the app_versions before startup, as well as causing some invalid keys and syntax, but couldn't get mouseless to fail when opening.

some other questions:
- were you able to reproduce the issue?
- do you have the config.yaml (and user_prefs.yaml) that were/are present when the failure happens?
- were you updating from v0.2.0 or v0.2.1?
- did you do any manually editing of the config, or user_prefs, or license files?

aside: there is an issue i need to fix with the app_version listed in the config (it doesn't update to the current version when the config is saved), thank you for pointing it out!  i don't think it causes any errors in program behavior right now, but i will fix it, of course.

thank you very much for your continued help on this!!

**Comment by twiggshub** [comment]
More than happy to answer with what I can. 

* As of now I can’t reproduce the problem and I can’t recover the files user.prefs.yaml, think I got them all confused, jumbled up and lost while trying to figure out what the fix was. It may have been a weird config, but as you seem to describe it shouldn’t be the reason why if you did all that.  If it helps, I did delete the config file during this entire process several times, and also before removing the data folder….  Same problem still occurred.  

Also I don’t remember if I mentioned this in the earlier comments but the only things that the config.yaml file that come stock with the one I had when the issue came were: a few added comments (though I doubt that did anything) and (maybe this is possible?) ordering the settings around comment-headers like “##subgrid settings” etc. Not sure this could be the reason though. Other than that, nothing done on any other file manually. 



As for updating, latest stable version (0.2.1) to the preview version.

**Comment by croian** [comment]
gotcha, thank you for the info!  so to confirm, order of events were:

- you opened mouseless, nothing happened (just the mouseless process)
- you deleted the config
- the issue with mouseless persisted even after restarting it
- you deleted the entire data folder
- you reinstalled
- you were then able to start mouseless successfully

if that's the case, that is definitely useful information, as it might rule out a corrupted config 

i'm not sure if there's much we can do to further diagnose -- what I probably have to do is just check all my file handling code, validation, etc at startup and making sure I'm handling errors properly.  my best guess is that there was a disk read error of some sort, or maybe a parse or validation error that i'm just not currently catching.

if you encounter the error again, please keep the files as they may help us diagnose! :)

if you can confirm (or deny) the order of events above, and let me know what macOS version your on, I would appreciate it!  then I think the next step is for me to dig into the code when I can, and we'll continue the conversation in the new issue i just made for this, #81

**Comment by croian** [comment]
one other thing @twiggshub, i know you said your files got all jumbled up and confused and lost -- but if you didn't actually delete them and can get them to me, they may help me pinpoint the bug much more quickly.  doesn't matter if you don't know which files are from when, i might be able to figure it out :)

zipping them and [emailing them](https://mouseless.click/support) to me is probably the best way

if you don't have them or can't find them, no worries :)

**Comment by vlnn** [comment]
> 1. i think a quick and easy solution to alleviate 1) is to simply have a special case for the globe/fn key, and allow it to pass through to the system while the overlay is up.  not a perfect solution, as some users use different shortcuts to switch layouts (i've been using ctrl+space while trying to fix this issue, because i don't have a fn key on my keyboard), and because, as you mention, we ultimately have issue 2) to contend with.

Totally agree: this is not the perfect solution. Moreover, someone may want to prefer globe/fn as a "modification button" as `right click modifier` or something — I would want for sure!

> 2. the ultimate solution, i think would indeed be to have the app automatically recognize layouts and map chars to keycodes and vice versa, without needing any configuration or effort by the user.  i think there are macos APIs for this, but it might take some extra legwork to get them working with mouseless, but I will investigate a path for doing so ASAP

Again, agree that this would be the ultimate solution. One small problem to consider though is that current configuration for `mouse action keys` is character-based, and it should either become key-code (which is may be confusing for the user, as there's no such thing as key-code "Й" for instance) or be mapped on the fly. Still, for me personally it's preferrable solution.

> 3. another solution in the meantime, would be to have an interface for the user enter all the keys for their different layouts, and map them to keycodes.  it wouldn't take too long, but still might annoy and confuse users as to why its necessary. but if that's what it takes to solve the issue soon it might be the best option for now.
>
> * just curious: you mention vim/vimperator/pentadactyl have some sort of configuration like this -- do they actually require the user to map characters to characters?  or do they use something like the approach i mentioned above?

For years vim/vimperator required this mapping and it was the first thing I've put into .vimrc. (see UA example here https://github.com/ipotseluev/configs/blob/8f57a36b11eb6fa721836f479747e124fb651c03/nvim/init.vim#L88 or russian-based example here https://dracoater.github.io/2020/03/vim-langmap). I guess now there are some prebuilt keymaps for that covering most user's needs. Still it's must have for people with custom layouts (imagine Dvorak for `EN` and ЙЦУКЕН for `UA`).
Mapping the characters is my "next preference" of a solution — maybe you could just provide several different mappings by default, and then make community create new ones? Just an idea.

> 4. finally, one of the most fragile alleviations would be to try to get mouseless to actually utilize the "automatically switch to document's input source" feature, but it might involve rapidly changing app/window focuses between mouseless and the previously focused window, which sounds very janky

I don't like this approach TBH as now for different focused windows user will have different popup (if I understand correctly that the popup will be automatically change the cell characters to the ones that can be input with current input mode). Still it's better than what we have right now.

> * just so i understand the problem space better though -- if i did get this approach to work, would that mostly solve (2) for you, or anyone, who has the "automatically switch" option enabled?

It will resolve the problem of not being able to use Mouseless, but will introduce quite some friction to "switch mentally". Best interfaces are reproducible and reliable — e.g. user expects same dialog for same actions, and in this case language will be different. 

Simply put, when I'm clicking mouse, I don't think about the language of my keyboard at the moment — I'm clicking mouse. When I'll click the Mouseless, I'd love not to think about the language as well — it should be either hardcoded or (even better) configurable.

Thank you @croian, as Mouseless.app is great piece of software already, and I hope it will be accessible for bigger community. I guess you'll find the right balance between the ease of use and generalization of the use cases! (keeping in mind the maintainability)

**Comment by croian** [comment]
> Thank you @croian, as Mouseless.app is great piece of software already, and I hope it will be accessible for bigger community. I guess you'll find the right balance between the ease of use and generalization of the use cases! (keeping in mind the maintainability)

thank you, @vlnn!  you've been a particularly excellent source of feedback and ideas.  mouseless will be an even better piece of software thanks to you!  🙏  and the others who've contributed of course!  🙏  i can sense a vibrant community beginning to form 🙌

> For years vim/vimperator required this mapping...

thank you for the linked examples, helps to see precisely how others have done it.  if I do an approach like this, i will make it as fast and user friendly as possible -- the user is guided to simply press all the keys for each layout, and mouseless does the rest behind the scenes.

also, quick aside -- I'm thinking allow user-defined exceptions for keys or key combos that should passthrough to the system might be a nice feature to have long-term.  would facilitate the Fn key for this particular use case, but I'm sure there are many other use cases!

i will look to implement solution 2 (no user config needed) or 3 (use enters keys for different layouts) in the next couple weeks!

**Comment by yecine06** [comment]
hey
i just found out about mouseless on "random youtubing" 

i am in french and testing with an azerty layout and even with the preview some characters are not working for me i think that it's for the keys that are on the third line of the keyboard:

wxcvbn

let me know what i can do to help you debug this

**Comment by croian** [comment]
NOTE: [v0.2.2-preview.1](https://github.com/croian/mouseless-issues/releases/tag/v0.2.2-preview.1) has been uploaded
